### ISSUE TITLE
Avatars look pixelated in some parts of UI, on high-density displays

### ISSUE THREAD

**Issue by gnprice** [zulip/zulip-mobile#4305]
(Originally diagnosed this at https://github.com/zulip/zulip-mobile/pull/4230#discussion_r521713861 , while reviewing the big avatar refactor in #4230 earlier today. But I've been noticing these symptoms for a long time.)

Repro:
* Take a device with a high-density display, like about 400ppi or higher.
  * Any flagship phone from the last few years will do; I see this on my Pixel 4 with its 444ppi display.
  * Or fire up an emulation of such a device, on a computer with a high-resolution monitor. This can make the effect easier to see, because the physical pixels are bigger.
* Optionally, to make the effect stronger: go to the device's display settings and turn up the display size to max. (E.g., Settings > Display > Display size > Larger, on stock Android 11.)
* Open the app and:
  * go to the PM conversations screen and look at the user avatars there;
  * or to "Create group" or "Search" from there, and look at the user avatars on the user list/search screen there;
  * or just look at your avatar in the lower-right corner from the main screen of the app;
  * or tap that one and look at your avatar in a large size at the top of the your-account screen.
* Expected: each avatar shows up nice and crisp.
* Actual: many avatars show up as some combination of pixelated, jagged, and blurry.
  * In short, they're low-res images which have been scaled up for display. The fact they're sometimes jagged, and aren't just blatantly pixelated or blurry, is presumably due to some sophisticated visual scaling algorithm that some layer in the stack has brought to bear to valiantly try to scale them up as cleanly as possible.

The exact symptoms are different between avatars that were uploaded to the Zulip server and avatars hosted on Gravatar; Gravatar will serve an avatar at any size, but Zulip provides avatars at just two discrete sizes, which can either aggravate this bug or mask it. They're also different between Gravatars on an org with hidden emails (like chat.zulip.org) vs. Gravatars on an org without hidden emails.

On a test account with an autogenerated Gravatar, here's what it looks like in that lower-right icon (as seen on an emulated Pixel 3, because using the emulator on my desktop makes cropped screenshots easy; and this is with default display settings):
![image](https://user-images.githubusercontent.com/28173/98883856-bc919b80-2443-11eb-9a83-c487ceb04a3c.png)
Yuck. That's pretty much how it looks on my actual Pixel 4, too; though it's a little less conspicuous there, just because the physical pixels are smaller.

Here's how it should look -- concretely, how it looks with a quick hack to fix this issue for just that widget:
![image](https://user-images.githubusercontent.com/28173/98884078-1c884200-2444-11eb-8833-eade4fde8ee5.png)
Much better!

Here's how my actual profile picture, which is a photo uploaded to the Zulip server, looks on the PMs screen:
![image](https://user-images.githubusercontent.com/28173/98884330-98828a00-2444-11eb-9315-405ec876a23f.png)
A little less blatant. The issue is mitigated a bit by that discrete sizing, and I think also the scaling-up algorithm may give better visual results on photos of people than on the highly polygonal autogenerated Gravatars.

But here's how it should look -- again, how it does look given a quick hack of a spot-fix:
![image](https://user-images.githubusercontent.com/28173/98884527-19418600-2445-11eb-8c8b-61a878f71f7d.png)
Distinctly better!

---

The basic issue is that our avatars code has confused [logical pixels vs. physical pixels](https://github.com/zulip/zulip-mobile/blob/d6254ad75ee0c327e37b382190f2dd041e858213/docs/background/layout.md). In a UI widget where we've laid out a space of N by N *logical* pixels to show an avatar, we fetch an image which contains only N by N pixels... and that space has a lot more than N by N actual physical pixels it needs to supply. On a modern phone, there will be several physical pixels to a logical pixel: usually about 2 or 3, sometimes more than 4.

For a concrete example:
* In that lower-right icon, the "size" is 24; we pass `size={24}` to `OwnAvatar`.
* That size is what we use for layout; so it's in *logical* pixels.
* The device has a lot more than 24 actual physical pixels in a span of 24 logical pixels. For example: on a Pixel 4, by default, the scaling factor is 2.75x, so there are 2.75 x 24 = 64 physical pixels in there.
* But in `OwnAvatar` we pass that `size` prop right through to `getAvatarFromUser`... which ends up looking for an avatar with that many *physical* pixels on each side.
  * For Zulip-hosted avatars, the issue gets masked here because the smallest avatar served is 100x100.
  * But for Gravatars, we get an actual 24x24 image. Blown up to the 64x64 physical pixels we need, it looks... like the screenshot above.

Or:
* On the PMs screen, each avatar has a "size" of 48; we pass `size={48}` to `UserAvatarWithPresence`.
* Again that's logical pixels; and on, for example, a Pixel 4 with default settings, that'd be 2.75 x 48 = 128 physical pixels.
* But if you trace down through the `getAvatarUrl` call in `UserAvatarWithPresence`, you find we end up looking for an avatar of 80x80 physical pixels. Scale that (or 100x100) up to 128x128, and it looks like what we see above.

---

For how to address this:
* Ideally we might ask the system what the logical-to-physical scaling factor is, and get an image of exactly the right physical size.
* For a good approximation, we can get an image with a (physical) size of 4x the logical size we'll be showing it at -- 4x is the density of [Android's `xxxhdpi` category for image assets](https://developer.android.com/training/multiscreen/screendensities), which is their highest.
* Concretely in the code:
  * `getAvatarUrl` and its friends (or `AvatarURL#get` after #4230) should make clear whether their "size" parameter is meant in logical or physical pixels.
    * Probably physical makes most sense -- that way that code stays focused on the intricacies of the different forms of avatar, and doesn't try to know things about the physical display of the device it's running on. That also coincides with how it currently behaves.
  * Then the call sites should arrange to pass something appropriate. Possibly just multiply the logical size (which they'll know, because they're UI code) by 4 -- except instead of `4` it should be something defined in one place with a descriptive name.
  * Then as future work we could try to identify the right scaling factor dynamically.

**Comment by abhi0504** [comment]
@gnprice are you working on this issue or can I take it

**Comment by chrisbobbe** [comment]
> For how to address this:
> 
> * Ideally we might ask the system what the logical-to-physical scaling factor is, and get an image of exactly the right physical size.

@gnprice, it looks like React Native has something called [`PixelRatio`](https://reactnative.dev/docs/0.62/pixelratio) for this! Do you see anything there other than `PixelRatio.get()` that you'd like to use?

**Comment by chrisbobbe** [comment]
https://github.com/zulip/zulip-mobile/pull/4230 aims to fix this along with its main goal. 🎉

**Comment by gnprice** [comment]
> it looks like React Native has something called [`PixelRatio`](https://reactnative.dev/docs/0.62/pixelratio) for this! Do you see anything there other than `PixelRatio.get()` that you'd like to use?

Great -- that looks like it should be exactly right!
